include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,317 @@
 // Licensed under the MIT License.
 
+#pragma once
+
 #include "core/framework/data_types.h"
 #include "core/framework/tensor_shape.h"
 #include "core/framework/tensor.h"
 
 namespace onnxruntime {
 
+class IDataTransfer;
+class DataTransferManager;
+
+/**
+ * @brief This is a Sparse Format enumeration representing bitflags
+ * 
+ * 
+ */
+enum class SparseFormatFlags : uint32_t {
+  kUndefined = 0x0U,        // For completeness
+  kCoo = 0x1U,              // 1-D or 2-D indices
+  kCsrc = 0x1U << 1,        // Both CSR(C)
+  kBlockSparse = 0x1U << 2  // as in OpenAI
+};
+
+std::ostream& operator<<(std::ostream&, SparseFormatFlags);
+
+inline SparseFormatFlags operator|(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return static_cast<SparseFormatFlags>(static_cast<uint32_t>(flags) | static_cast<uint32_t>(flag));
+}
+
+inline SparseFormatFlags operator&(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return static_cast<SparseFormatFlags>(static_cast<uint32_t>(flags) & static_cast<uint32_t>(flag));
+}
+
+inline SparseFormatFlags TurnOff(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return static_cast<SparseFormatFlags>(static_cast<uint32_t>(flags) & ~static_cast<uint32_t>(flag));
+}
+
+inline bool IsSet(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return SparseFormatFlags::kUndefined != (flags & flag);
+}
+
+inline SparseFormatFlags Set(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return flags | flag;
+}
+
+/**
+ * @brief This is a base class for virtual sparse format representation
+ * 
+ */
+class SparseRep {
+ protected:
+  SparseRep(MLDataType data_type, const TensorShape& values_shape, int64_t total_buffer_size, const AllocatorPtr& allocator);
+  SparseRep(MLDataType data_type, const TensorShape& values_shape, void* values, const OrtMemoryInfo& mem_info);
+
+  void* IndicesStart(int64_t align) {
+    return reinterpret_cast<uint8_t*>(values_.MutableDataRaw()) + Roundup(values_.SizeInBytes(), align);
+  }
+
+ public:
+  virtual ~SparseRep();
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(SparseRep);
+
+  const Tensor& Values() const noexcept {
+    return values_;
+  }
+
+  Tensor& MutableValues() noexcept {
+    return values_;
+  }
+
+  bool OwnBuffer() const noexcept {
+    return allocator_ != nullptr;
+  }
+
+  const OrtMemoryInfo& Location() const {
+    return values_.Location();
+  }
+
+  // Round up size to a multiple of align.
+  // Example:
+  //   Roundup(13, 5)   => 15
+  //   Roundup(201, 16) => 208
+  static int64_t Roundup(int64_t size, int64_t align) {
+    return ((size + align - 1) / align) * align;
+  }
+
+  /// <summary>
+  /// Calculate required buffer size. We will place indices
+  /// after data and make sure indices start at int64_t aligned place
+  /// </summary>
+  /// <returns></returns>
+  static int64_t CalculateRequiredBufferSize(int64_t data_size, int64_t indices_size, int64_t align) {
```

Comment:
>CalculateRequiredBufferSize [](http://example.com/codeflow?start=17&length=27)

should we use SafeInt for computing buffer sizes?

Reply:
we MUST use it for any calculation that determines the size of a memory allocation.